bd-client-common/src/safe_file_cache.rs

```diff
@@ -0,0 +1,171 @@
+    // allow a client that received bad configuration to eventually recover, avoiding an infinite
+    // crash loop.
+
+    // If either of the files don't exist, we're not going to try to load the config and we'll wipe
+    // out the other file if it's there. This could handle naturally if the system shuts down in the
+    // middle of caching config.
+    if !tokio::fs::try_exists(self.retry_count_file())
+      .await
+      .is_ok_and(|e| e)
+      || !tokio::fs::try_exists(self.protobuf_file())
+        .await
+        .is_ok_and(|e| e)
+    {
+      log::debug!("cached retry count or config not found, resetting cache");
+      return Ok((true, None));
+    }
+
+    // If the retry count file contains invalid data we defensively bail on reading the cached
+    // value. If we were to treat an empty file as count=0 we could theoretically find ourselves in
+    // a loop where the file is not properly updated.
+    let Ok(retry_count) = Self::parse_retry_count(
+      &tokio::fs::read(&self.retry_count_file())
+        .await
+        .map_err(|e| anyhow!("an io error occurred: {e}"))?,
+    ) else {
+      return Ok((true, None));
+    };
+
+    log::debug!("loaded file at retry count {retry_count}");
+
+    // Update the retry count before we apply the config. If this fails, we bail out (which will
+    // attempt to clear the cache directory) and disable caching. We do this because being unable
+    // to update the retry count may result in us getting stuck processing what we think is retry
+    // 0 over and over again.
+    self.persist_cache_load_retry_count(retry_count + 1).await?;
+
+    // TODO(snowp): Should we read this from runtime as well? It would make it possible for a bad
+    // runtime config to accidentally set this really high, but right now this is not
+    // configurable at all.
+    if retry_count > 5 {
+      // Note that eventually if the client is killed this is going to kick in and delete cached
+      // config. This is OK since we are still covered by the kill file, and ultimately we would
+      // like the client to come up and get fresh config anyway.
+      return Ok((true, None));
+    }
+
+    // TODO(snowp): Add stats for how often the read fails beyond ENOENT.
+    // TODO(snowp): When we add a callback, only trigger it if we successfully loaded the config.
+    let bytes = tokio::fs::read(&self.protobuf_file()).await?;
+    let protobuf: T = read_compressed_protobuf(&bytes)?;
+
+    Ok((false, Some(protobuf)))
+  }
+
+  fn parse_retry_count(data: &[u8]) -> anyhow::Result<u32> {
+    Ok(std::str::from_utf8(data)?.parse()?)
+  }
+
+  pub async fn cache_update(&self, protobuf: &T) {
+    if let Err(e) =
+      tokio::fs::write(&self.protobuf_file(), write_compressed_protobuf(protobuf)).await
+    {
+      log::debug!("failed to write cached config: {e}");
+    }
+
+    // Failing here is fine, worst case we'll use an old retry count or leave is missing, which
```

Comment:
```suggestion
    // Failing here is fine, worst case we'll use an old retry count or leave it missing, which
```
I think? oddly phrased comment, probably one of mine

Reply:
Will fix in the next PR just copied this.